opencog/reasoning/RuleEngine/rule-engine-src/pln/FCMemory.cc

```diff
@@ -22,107 +22,129 @@
-	target_list_.push_back(cur_target_);
-}
-HandleSeq FCMemory::get_target_list(void) {
-	return target_list_;
-}
-bool FCMemory::is_search_in_af(void) {
-	return search_in_af_;
-}
-Rule* FCMemory::get_cur_rule(void) {
-	return cur_rule_;
-}
-void FCMemory::set_cur_rule(Rule* r) {
-	cur_rule_ = r;
-}
-void FCMemory::add_rules_product(int iteration, HandleSeq product) {
-	for (Handle p : product) {
-		Inference inf;
-		inf.iter_step = iteration;
-		inf.applied_rule = cur_rule_;
-		inf.inf_product.push_back(p);
-		inf_history_.push_back(inf);
-	}
+vector<Rule*> FCMemory::get_rules(void)
+{
+    return rules_;
+}
+void FCMemory::set_rules(vector<Rule*> rules)
+{
+    rules_ = rules;
+}
+void FCMemory::set_target(Handle target)
+{
+    cur_target_ = target;
+    target_list_.push_back(cur_target_);
+}
+HandleSeq FCMemory::get_target_list(void)
+{
+    return target_list_;
+}
+HandleSeq FCMemory::get_premise_list(void)
+{
+    return premise_list_;
+}
+bool FCMemory::is_search_in_af(void)
+{
+    return search_in_af_;
+}
+Rule* FCMemory::get_cur_rule(void)
+{
+    return cur_rule_;
+}
+void FCMemory::set_cur_rule(Rule* r)
+{
+    cur_rule_ = r;
+}
+void FCMemory::add_rules_product(int iteration, HandleSeq product)
+{
+    for (Handle p : product) {
+        Inference inf;
+        inf.iter_step = iteration;
+        inf.applied_rule = cur_rule_;
+        inf.inf_product.push_back(p);
+        inf_history_.push_back(inf);
+    }
 }
 void FCMemory::add_inference(int iter_step, HandleSeq product,
-		HandleSeq matched_nodes) {
-	Inference inf;
-	inf.applied_rule = cur_rule_;
-	inf.iter_step = iter_step;
-	for (Handle p : product)
-		inf.inf_product.push_back(p);
-	for (Handle mn : matched_nodes)
-		inf.matched_nodes.push_back(mn);
-	inf_history_.push_back(inf);
-}
-Handle FCMemory::get_cur_target(void) {
-	return cur_target_;
-}
-
-bool FCMemory::isin_target_list(Handle h) {
-	for (Handle hi : target_list_) {
-		if (hi.value() == h.value())
-			return true;
-		//recursive lookup
-		else if (LinkCast(hi)) {
-			HandleSeqSeq hseqs;
-			hseqs.push_back(as_->getOutgoing(hi));
-			do {
-				HandleSeq iset = hseqs[hseqs.size() - 1];
-				hseqs.pop_back();
-				for (Handle i : iset) {
-					if (i.value() == h.value())
-						return true;
-					else if (LinkCast(i))
-						hseqs.push_back(as_->getOutgoing(i));
-				}
-			} while (not hseqs.empty());
-		}
-	}
-	return false;
+                             HandleSeq matched_nodes)
+{
+    Inference inf;
+    inf.applied_rule = cur_rule_;
+    inf.iter_step = iter_step;
+    for (Handle p : product)
+        inf.inf_product.push_back(p);
+    for (Handle mn : matched_nodes)
+        inf.matched_nodes.push_back(mn);
+    inf_history_.push_back(inf);
+}
+Handle FCMemory::get_cur_target(void)
+{
+    return cur_target_;
+}
+bool FCMemory::isin_target_list(Handle h)
+{
+    return (find(target_list_.begin(), target_list_.end(), h) != target_list_.end());
+}
+bool FCMemory::isin_premise_list(Handle h)
```

Comment:
Hopefully you will start documenting your code soon.... I have not idea what is a "target" and what is a "premise" and why you are renaming it in some place and not others.